contracts/schemes/TokenTrade.sol

```diff
@@ -0,0 +1,159 @@
+     */
+    function initialize(
+        Avatar _avatar,
+        IntVoteInterface _votingMachine,
+        uint256[11] calldata _votingParams,
+        address _voteOnBehalf,
+        bytes32 _voteParamsHash
+    )
+    external
+    {
+        super._initializeGovernance(_avatar, _votingMachine, _voteParamsHash, _votingParams, _voteOnBehalf);
+    }
+
+    /**
+    * @dev execution of proposals, can only be called by the voting machine in which the vote is held.
+    * @param _proposalId the ID of the voting in the voting machine
+    * @param _decision a parameter of the voting result, 1 yes and 2 is no.
+    * @return bool success
+    */
+    function executeProposal(bytes32 _proposalId, int256 _decision)
+    external
+    onlyVotingMachine(_proposalId)
+    override
+    returns(bool) {
+        Proposal memory proposal = proposals[_proposalId];
+        delete proposals[_proposalId];
+        if (_decision == 1 && proposal.receiveToken.balanceOf(address(avatar)) >= proposal.receiveTokenAmount) {
+            proposal.sendToken.safeTransfer(address(avatar), proposal.sendTokenAmount);
+            require(
+                Controller(avatar.owner()).externalTokenTransfer(
+                    proposal.receiveToken, proposal.beneficiary, proposal.receiveTokenAmount
+                ), "failed to transfer tokens from the DAO"
+            );
+
+            emit TokenTradeProposalExecuted(
+                address(avatar),
+                _proposalId,
+                proposal.beneficiary,
+                proposal.sendToken,
+                proposal.sendTokenAmount,
+                proposal.receiveToken,
+                proposal.receiveTokenAmount
+            );
+        } else {
+            proposal.sendToken.safeTransfer(address(proposal.beneficiary), proposal.sendTokenAmount);
+        }
+
+        emit ProposalExecuted(address(avatar), _proposalId, _decision);
+        return true;
+    }
+
+    /**
+    * @dev propose to trade tokens with the DAO
+    * @param _sendToken token the proposer suggests to send to the DAO
+    * @param _sendTokenAmount token amount the proposer suggests to send to the DAO
+    * @param _receiveToken token the proposer asks to receive from the DAO
+    * @param _receiveTokenAmount token amount the proposer asks to receive from the DAO
+    * @param _descriptionHash proposal description hash
+    * @return an id which represents the proposal
+    */
+    function proposeTokenTrade(
+        IERC20 _sendToken,
+        uint256 _sendTokenAmount,
+        IERC20 _receiveToken,
+        uint256 _receiveTokenAmount,
+        string memory _descriptionHash
+    )
+    public
+    returns(bytes32)
+    {
+        require(
+            address(_sendToken) != address(0) && address(_receiveToken) != address(0),
+            "Token address must not be null"
+        );
+        require(_sendTokenAmount > 0 && _receiveTokenAmount > 0, "Token amount must be greater than 0");
+
```

Comment:
I think we need a way to `lock` the recieve token in the scheme.. so there will not be a case where the proposal execute and the dao cannot pay .
so maybe upon proposal the scheme will ask the receive token amount from the avatar.. ,if not enough exist will revert.
if proposal is accepted the user will redeem (from the scheme) .
if proposal is rejected the allocated token will be sent back to the avatar

Reply:
This is bad because an attacker can open dummy proposal and lock the avatar's balance. The way I handled that case is that if upon execution the DAO doesn't have enough funds, the proposal will be canceled as if it was rejected, and the user will get his tokens back.